src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.RawSocket.cs

```diff
@@ -0,0 +1,376 @@
+
+                long elapsed;
+                Stopwatch sw = Stopwatch.StartNew();
+                // Read from the socket in a loop. We may receive messages that are not echo replies, or that are not in response
+                // to the echo request we just sent. We need to filter such messages out, and continue reading until our timeout.
+                // For example, when pinging the local host, we need to filter out our own echo requests that the socket reads.
+                while ((elapsed = sw.ElapsedMilliseconds) < timeout)
+                {
+                    int bytesReceived;
+                    try
+                    {
+                        bytesReceived = socket.ReceiveFrom(receiveBuffer, SocketFlags.None, ref socketConfig.EndPoint);
+                    }
+                    catch (SocketException ex) when (ex.SocketErrorCode == SocketError.TimedOut)
+                    {
+                        return CreateTimedOutPingReply();
+                    }
+
+                    if (bytesReceived - ipHeaderLength < IcmpHeaderLengthInBytes)
+                    {
+                        continue; // Not enough bytes to reconstruct IP header + ICMP header.
+                    }
+
+                    if (TryGetPingReply(socketConfig, receiveBuffer, bytesReceived, sw, ref ipHeaderLength, out PingReply? reply))
+                    {
+                        return reply;
+                    }
+                }
+
+                // We have exceeded our timeout duration, and no reply has been received.
+                return CreateTimedOutPingReply();
+            }
+        }
+
+        private async Task<PingReply> SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, byte[] buffer, int timeout, PingOptions? options)
+        {
+            SocketConfig socketConfig = GetSocketConfig(address, buffer, timeout, options);
+            using (Socket socket = GetRawSocket(socketConfig))
+            {
+                int ipHeaderLength = socketConfig.IsIpv4 ? MinIpHeaderLengthInBytes : 0;
+
+                await socket.SendToAsync(
+                    new ArraySegment<byte>(socketConfig.SendBuffer),
+                    SocketFlags.None, socketConfig.EndPoint)
+                    .ConfigureAwait(false);
+
+                byte[] receiveBuffer = new byte[MaxIpHeaderLengthInBytes + IcmpHeaderLengthInBytes + buffer.Length];
+
+                long elapsed;
+                Stopwatch sw = Stopwatch.StartNew();
+                // Read from the socket in a loop. We may receive messages that are not echo replies, or that are not in response
+                // to the echo request we just sent. We need to filter such messages out, and continue reading until our timeout.
+                // For example, when pinging the local host, we need to filter out our own echo requests that the socket reads.
+                while ((elapsed = sw.ElapsedMilliseconds) < timeout)
+                {
```

Comment:
Did you consider using overloads with cancellationtoken instead of manual StopWatch?

Reply:
I tried to rewrite it to use `CancellationToken`s but in the end the stopwatch is also used to measure the ping time so it's needed anyway.

Reply:
You could possibly use `Environment.TickCount64` directly.